### PR TITLE
Resolve #249 - Entropy 0.3 - the rest

### LINKED ISSUE #249: Grammar extension strategy for entropy
Entropy 0.2, which I'm currently implementing, supports extending floats, strings, etc. but does not support adding interfaces or string enums or even boolean values that do not appear in the dictionary in a specific context.

The following is a simple mechanism for extensibility. It builds upon the existing extension mechanism (prelude / content streams) and does not attempt any kind of optimization, as the idea is that we will update the dictionary every so often to take into account new constructs that have become popular around the web.

# Dictionary generation

When we generate a probability table, we add in every context a single symbol (the "extension value") with the minimal probability.

# Encoding

When we attempt to encode a string enum/interface name/boolean value that does not appear in the probability table in this context, we:
- encode the *extension value*;
- fetch the index of the value from the the prelude dictionary for string enums (respectively interface names, boolean values)
- if there is no index value, append it to the dictionary, creating a new index;
- append the index to the content stream for string enums (respectively interface names, boolean values)

# Decoding

When we extract the *extension symbol* in a context:
- read the index from the content stream for string enums (respectively interface names, boolean values);
- get the value at said index from the prelude dictionary of string enums (respectively interface names, boolean values).

If there is no index or no value, it's a syntax error.

### PR DESCRIPTION
**Pull request by Yoric**
This finishes connecting all the bits and pieces of Entropy 0.3.

With this PR, the compressor can finally compress data whose grammar constructions do not show up in the samples. In particular, this means that we can (poorly) compress without dictionary.

### CODE REVIEW

File: crates/binjs_shared/src/ast.rs [hunk 1]
```diff
@@ -261,3 +270,4 @@ pub trait Node: downcast_rs::Downcast {
     fn name(&self) -> &'static str;
 }
 impl_downcast!(Node);
+
```
Comment (arai-a) [inline crates/binjs_shared/src/ast.rs]: remove this maybe?

Comment (Yoric) [reply crates/binjs_shared/src/ast.rs]: The newline?

Comment (arai-a) [reply crates/binjs_shared/src/ast.rs]: yes, extra newline at the end of file.
if the diff is shown correctly, now the file ends with `\n\n`, but it should end with `\n`

Comment (Yoric) [reply crates/binjs_shared/src/ast.rs]: Actually, this gets rid of a missing newline, so I think I'll keep it :)

**Pull request merged**